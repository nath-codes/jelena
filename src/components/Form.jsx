import React, { Component } from "react";
import styled from "styled-components";
import Label from "./Label";
import Input from "./Input";
import Fieldset from "./Fieldset";
import FormHeading from "./FormHeading";
import Button from "./Button";
import { FormList, FormListItem } from "./FormList";
import sendMail from "../services/send-email";
import Text from "./Text";
import spacer from "../mixins/spacer";

const StyledForm = styled.form`
  display: flex;
  flex-direction: column;
  min-width: 50vw;

  .message {
    margin: ${spacer(1)} 0;
  }
`;

class Form extends Component {
  constructor(props) {
    super(props);

    this.state = {
      attending: "Yes",
      dessert: "",
      dietaryRequirements: "",
      fullName: "",
      main: "",
      sending: false,
      sent: false,
      starter: ""
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    const { target } = event;
    const value = target.type === "checkbox" ? target.checked : target.value;
    const name = target.name;
    this.setState({ [name]: value });
  }

  async handleSubmit(event) {
    event.preventDefault();
    this.setState({ sending: true });

    const response = await sendMail(this.state);
    this.handleResponse(response);
  }

  handleResponse(response) {
    const { status } = response;

    if (status === 200) {
      this.setState({
        dessert: "",
        dietaryRequirements: "",
        fullName: "",
        main: "",
        sending: false,
        sent: true,
        starter: ""
      });
    }
  }

  render() {
    return (
      <StyledForm onSubmit={this.handleSubmit}>
        <Fieldset>
          <Input
            name="fullName"
            placeholder="Enter your name here"
            type="text"
            required="required"
            value={this.state.fullName}
            onChange={this.handleChange}
          />
        </Fieldset>

        <FormHeading>Attending</FormHeading>

        <Fieldset>
          <FormList inline>
            <FormListItem orphan>
              <Label>
                <Input
                  name="attending"
                  type="radio"
                  value="Yes"
                  onChange={this.handleChange}
                  defaultChecked
                />Yes
              </Label>
            </FormListItem>
            <FormListItem>
              <Label>
                <Input
                  name="attending"
                  type="radio"
                  value="No"
                  onChange={this.handleChange}
                />
                No
              </Label>
            </FormListItem>
          </FormList>
        </Fieldset>

        {this.state.attending === "Yes" && (
          <div>
            <FormHeading>Starter</FormHeading>
            <Fieldset>
              <FormList>
                <FormListItem>
                  <Label>
                    <Input
                      type="radio"
                      name="starter"
                      value="Grilled pegetable platter (vegan)"
                      onChange={this.handleChange}
                    />
                    Grilled vegetable platter <span>(vegan)</span>
                  </Label>
                </FormListItem>
                <FormListItem>
                  <Label>
                    <Input
                      type="radio"
                      name="starter"
                      value="Ham hock terrine (gf)"
                      onChange={this.handleChange}
                    />
                    Ham hock terrine <span>(gf)</span>
                  </Label>
                </FormListItem>
              </FormList>
            </Fieldset>

            <FormHeading>Main</FormHeading>
            <Fieldset>
              <FormList>
                <FormListItem>
                  <Label>
                    <Input
                      type="radio"
                      name="main"
                      value="Roast lamb with dauphinoise potatoes (gf)"
                      onChange={this.handleChange}
                    />
                    Roast lamb with dauphinoise potatoes <span>(gf)</span>
                  </Label>
                </FormListItem>
                <FormListItem>
                  <Label>
                    <Input
                      type="radio"
                      name="main"
                      value="Vegetable tagine (vegan)"
                      onChange={this.handleChange}
                    />
                    Vegetable tagine <span>(vegan)</span>
                  </Label>
                </FormListItem>
              </FormList>
            </Fieldset>

            <FormHeading>Desserts</FormHeading>
            <Fieldset>
              <FormList>
                <FormListItem>
                  <Label>
                    <Input
                      type="radio"
                      name="dessert"
                      value="Lemon tart (vegan)"
                      onChange={this.handleChange}
                    />
                    Lemon tart <span>(vegan)</span>
                  </Label>
                </FormListItem>
                <FormListItem>
                  <Label>
                    <Input
                      type="radio"
                      name="dessert"
                      value="Brandy tart (gf)"
                      onChange={this.handleChange}
                    />
                    Brandy tart <span>(gf)</span>
                  </Label>
                </FormListItem>
              </FormList>
            </Fieldset>

            <Fieldset>
              <Input
                type="text"
                name="dietaryRequirements"
                placeholder="Any dietary requirements?"
                value={this.state.dietaryRequirements}
                onChange={this.handleChange}
              />
            </Fieldset>
          </div>
        )}

        <Button active={this.state.sending} type="submit">
          Submit
        </Button>

        {this.state.sent && (
          <Text className="message">Your RSVP has been recieved.</Text>
        )}
      </StyledForm>
    );
  }
}

export default Form;
