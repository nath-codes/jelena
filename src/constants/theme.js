export default {
  fonts: {
    heading: "bebas-neue",
    body: "montserrat"
  },
  colors: {
    delftBlue: "#1b56c5",
  }
};
