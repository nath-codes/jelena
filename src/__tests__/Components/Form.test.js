import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import "jest-styled-components";
import { ThemeProvider } from "styled-components";
import Form from "../../components/Form";
import theme from "../../constants/theme";

jest.mock("../../services/send-email", () => () => Promise.resolve());

describe("<Form />", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={theme}>
          <Form />
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("displays meal choices when guest is attending ", () => {
    const wrapper = shallow(<Form />);
    wrapper.setState({ attending: "Yes" });

    expect(wrapper).toMatchSnapshot();
  });

  it("displays success message when form is submitted", () => {
    const wrapper = shallow(<Form />);
    wrapper.setState({ sent: true });

    expect(wrapper).toMatchSnapshot();
  });

  describe("updating state", () => {
    it("should set checkbox inputs to state", () => {
      const wrapper = shallow(<Form />);
      const instance = wrapper.instance();
      const checked = "foo";
      const property = "dessert";

      const eventMock = {
        target: {
          type: "checkbox",
          name: property,
          checked
        }
      };

      instance.handleChange(eventMock);

      expect(wrapper.state(property)).toEqual(checked);
    });

    it("should set text inputs to state", () => {
      const wrapper = shallow(<Form />);
      const instance = wrapper.instance();
      const value = "foo";
      const property = "name";

      const eventMock = {
        target: {
          type: "whatever",
          name: property,
          value
        }
      };

      instance.handleChange(eventMock);

      expect(wrapper.state(property)).toEqual(value);
    });
  });

  describe("submissions", () => {
    const completedState = {
      attending: "No",
      dessert: "foo",
      dietaryRequirements: "foo",
      fullName: "foo",
      main: "foo",
      sending: false,
      sent: true,
      starter: "foo"
    };

    const resetState = {
      attending: "No",
      dessert: "",
      dietaryRequirements: "",
      fullName: "",
      main: "",
      sending: false,
      sent: true,
      starter: ""
    };

    it("should submit the form", async () => {
      const wrapper = shallow(<Form />);
      const instance = wrapper.instance();
      const handleResponseMock = jest.fn();
      const mockEvent = {
        preventDefault: () => {}
      };

      instance.handleResponse = handleResponseMock;
      await instance.handleSubmit(mockEvent);
      expect(handleResponseMock).toHaveBeenCalled();
    });

    it("should clear form state on success", () => {
      const wrapper = shallow(<Form />);
      const instance = wrapper.instance();
      const response = { status: 200 };

      wrapper.setState(completedState);
      instance.handleResponse(response);

      expect(wrapper.state()).toEqual(resetState);
    });

    it("should not clear form state on error", () => {
      const wrapper = shallow(<Form />);
      const instance = wrapper.instance();
      const response = { status: 401 };

      wrapper.setState(completedState);
      instance.handleResponse(response);

      expect(wrapper.state()).toEqual(completedState);
    });
  });
});
