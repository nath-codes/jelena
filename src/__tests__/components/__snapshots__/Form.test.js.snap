// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> displays meal choices when guest is attending  1`] = `
<styled.form
  onSubmit={[Function]}
>
  <styled.div>
    <styled.input
      name="fullName"
      onChange={[Function]}
      placeholder="Enter your name here"
      required="required"
      type="text"
      value=""
    />
  </styled.div>
  <styled.h3>
    Attending
  </styled.h3>
  <styled.div>
    <styled.ul
      inline={true}
    >
      <styled.li
        orphan={true}
      >
        <styled.label>
          <styled.input
            name="attending"
            onChange={[Function]}
            type="radio"
            value="Yes"
          />
          Yes
        </styled.label>
      </styled.li>
      <styled.li>
        <styled.label>
          <styled.input
            name="attending"
            onChange={[Function]}
            type="radio"
            value="No"
          />
          No
        </styled.label>
      </styled.li>
    </styled.ul>
  </styled.div>
  <div>
    <styled.h3>
      Starter
    </styled.h3>
    <styled.div>
      <styled.ul>
        <styled.li>
          <styled.label>
            <styled.input
              checked={false}
              name="starter"
              onChange={[Function]}
              required="required"
              type="radio"
              value="Grilled vegetable platter (vegan)"
            />
            Grilled vegetable platter 
            <span>
              (vegan)
            </span>
          </styled.label>
        </styled.li>
        <styled.li>
          <styled.label>
            <styled.input
              checked={false}
              name="starter"
              onChange={[Function]}
              type="radio"
              value="Ham hock terrine (gf)"
            />
            Ham hock terrine 
            <span>
              (gf)
            </span>
          </styled.label>
        </styled.li>
      </styled.ul>
    </styled.div>
    <styled.h3>
      Main
    </styled.h3>
    <styled.div>
      <styled.ul>
        <styled.li>
          <styled.label>
            <styled.input
              checked={false}
              name="main"
              onChange={[Function]}
              required="required"
              type="radio"
              value="Roast lamb with dauphinoise potatoes (gf)"
            />
            Roast lamb with dauphinoise potatoes 
            <span>
              (gf)
            </span>
          </styled.label>
        </styled.li>
        <styled.li>
          <styled.label>
            <styled.input
              checked={false}
              name="main"
              onChange={[Function]}
              type="radio"
              value="Vegetable tagine (vegan)"
            />
            Vegetable tagine 
            <span>
              (vegan)
            </span>
          </styled.label>
        </styled.li>
      </styled.ul>
    </styled.div>
    <styled.h3>
      Desserts
    </styled.h3>
    <styled.div>
      <styled.ul>
        <styled.li>
          <styled.label>
            <styled.input
              checked={false}
              name="dessert"
              onChange={[Function]}
              required="required"
              type="radio"
              value="Lemon tart"
            />
            Lemon tart
          </styled.label>
        </styled.li>
        <styled.li>
          <styled.label>
            <styled.input
              checked={false}
              name="dessert"
              onChange={[Function]}
              type="radio"
              value="Brandy tart (gf) (vegan)"
            />
            Brandy tart 
            <span>
              (gf) (vegan)
            </span>
          </styled.label>
        </styled.li>
      </styled.ul>
    </styled.div>
    <styled.div>
      <styled.input
        name="dietaryRequirements"
        onChange={[Function]}
        placeholder="Any dietary requirements?"
        type="text"
        value=""
      />
    </styled.div>
  </div>
  <styled.button
    active={false}
    type="submit"
  >
    Submit
  </styled.button>
</styled.form>
`;

exports[`<Form /> displays success message when form is submitted 1`] = `
<styled.form
  onSubmit={[Function]}
>
  <styled.div>
    <styled.input
      name="fullName"
      onChange={[Function]}
      placeholder="Enter your name here"
      required="required"
      type="text"
      value=""
    />
  </styled.div>
  <styled.h3>
    Attending
  </styled.h3>
  <styled.div>
    <styled.ul
      inline={true}
    >
      <styled.li
        orphan={true}
      >
        <styled.label>
          <styled.input
            name="attending"
            onChange={[Function]}
            type="radio"
            value="Yes"
          />
          Yes
        </styled.label>
      </styled.li>
      <styled.li>
        <styled.label>
          <styled.input
            name="attending"
            onChange={[Function]}
            type="radio"
            value="No"
          />
          No
        </styled.label>
      </styled.li>
    </styled.ul>
  </styled.div>
  <styled.button
    active={false}
    type="submit"
  >
    Submit
  </styled.button>
  <styled.div
    className="message"
  >
    Your RSVP has been recieved.
  </styled.div>
</styled.form>
`;

exports[`<Form /> renders correctly 1`] = `
.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: montserrat;
  font-size: 0.8999999999999999rem;
  width: 100%;
}

.c6 > span {
  font-size: 12px;
  margin-left: auto;
  padding-left: 6px;
}

.c2 {
  border: solid 3px #1b56c5;
  border-radius: 3px;
  font-family: montserrat;
  font-size: 0.8999999999999999rem;
  height: 2.4rem;
  padding: 0 0 0 0.6rem;
  width: 100%;
}

.c7 {
  margin-right: 0.6rem;
}

.c1 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: auto;
  margin: 0 0 2.4rem;
  padding: 0;
}

.c1 [object Object]::after {
  clear: both;
  content: "";
  display: table;
}

.c3 {
  color: #1b56c5;
  font-family: bebas-neue;
  font-size: 1.7999999999999998rem;
  line-height: 2.4rem;
  margin: 0 0 1.2rem;
}

.c9 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  background: #1b56c5;
  border: 0;
  color: #fff;
  cursor: pointer;
  font-family: bebas-neue;
  font-size: 1.2rem;
  line-height: 1;
  padding-bottom: 0.6rem;
  padding-top: 0.6rem;
  padding-left: 2.4rem;
  padding-right: 2.4rem;
  position: relative;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
}

.c4 {
  list-style-type: none;
  padding: 0;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c4 li:nth-child(2) {
  margin-left: 1.2rem;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 1.2rem;
  margin-bottom: 0.6rem;
  margin-bottom: 0;
}

.c5:last-child {
  margin-bottom: 0;
}

.c8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 1.2rem;
  margin-bottom: 0.6rem;
}

.c8:last-child {
  margin-bottom: 0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 28.799999999999997rem;
  width: 100%;
}

.c0 .message {
  margin: 1.2rem 0;
}

@media (min-width:576px) {
  .c6 > span {
    margin-left: 0;
  }
}

@media (min-width:768px) {
  .c6 > span {
    font-size: 14px;
  }
}

<form
  className="c0"
  onSubmit={[Function]}
>
  <div
    className="c1"
  >
    <input
      className="c2"
      name="fullName"
      onChange={[Function]}
      placeholder="Enter your name here"
      required="required"
      type="text"
      value=""
    />
  </div>
  <h3
    className="c3"
  >
    Attending
  </h3>
  <div
    className="c1"
  >
    <ul
      className="c4"
    >
      <li
        className="c5"
      >
        <label
          className="c6"
        >
          <input
            className="c7"
            name="attending"
            onChange={[Function]}
            type="radio"
            value="Yes"
          />
          Yes
        </label>
      </li>
      <li
        className="c8"
      >
        <label
          className="c6"
        >
          <input
            className="c7"
            name="attending"
            onChange={[Function]}
            type="radio"
            value="No"
          />
          No
        </label>
      </li>
    </ul>
  </div>
  <button
    className="c9"
    type="submit"
  >
    Submit
  </button>
</form>
`;
